reconcile/openshift_namespace_labels.py

```diff
@@ -0,0 +1,395 @@
+        """Get the registered errors for the given cluster / namespace.
+        Defaults to []"""
+        return self._errors.setdefault(cluster, {}).setdefault(namespace, [])
+
+    def add_error(self, cluster: str, namespace: str, err: str):
+        """Add an error to the given cluster / namespace"""
+        self.errors(cluster, namespace).append(err)
+
+    def has_any_error(self) -> bool:
+        """Checks if the given cluster / namespace has any error registered"""
+        return any(e[2] for e in self.iter_errors())
+
+    def iter_errors(self) -> Generator[Tuple[str, str, List[str]], None, None]:
+        """yields (cluster, namespace, errors) items"""
+        for cluster, namespaces in self._errors.items():
+            for namespace, errors in namespaces.items():
+                if errors:
+                    yield cluster, namespace, errors
+
+    def _ns(self, cluster: str, namespace: str) -> Types:
+        return self._inv.setdefault(cluster, {}).setdefault(namespace, {})
+
+    def get(self, cluster: str, namespace: str, type: str,
+            default: LabelsOrKeys = None) -> LabelsOrKeys:
+        """Get the labels or keys for the given cluster / namespace / type"""
+        return self._inv.get(cluster, {}).get(namespace, {}).get(type,
+                                                                 default)
+
+    def setdefault(self, cluster: str, namespace: str, type: str,
+                   default: LabelsOrKeys = None) -> LabelsOrKeys:
+        """Get the labels or keys for the given cluster / namespace / type,
+        setting it to default if it does not exists"""
+        with self._lock:
+            return self._ns(cluster, namespace).setdefault(type, default)
+
+    def set(self, cluster: str, namespace: str,
+            type: str, labels: Labels) -> Labels:
+        """Sets the given cluster / namespace / type to 'labels'"""
+        with self._lock:
+            self._ns(cluster, namespace)[type] = labels
+            return labels
+
+    def delete(self, cluster: str, namespace: str):
+        """Delete the given cluster / namespace from the inventory"""
+        with self._lock:
+            self._inv.get(cluster, {}).pop(namespace, None)
+
+    def __iter__(self) -> Generator[Tuple[str, str, Types], None, None]:
+        """Makes the inventory iterable by yielding (cluster, namespace, types)
+        items. Types here is a dict of {type: labelsOrKeys} """
+        for cluster, namespaces in self._inv.items():
+            for namespace, types in namespaces.items():
+                yield cluster, namespace, types
+
+    def update_managed_keys(self, cluster: str, namespace: str, key: str):
+        """
+        Add or remove a key from the managed key list.
+        This actually handles a copy of the managed keys dict and updates it.
+        If the key was managed, it will get removed.
+        If it was not, it will get added
+        """
+        managed = self.get(cluster, namespace, MANAGED, [])
```

Comment:
So, I thought about this more... what about this case:

```
inv = {'cluster': {'namespace': {'managed': None}}}
>>> type(inv.get('cluster').get('namespace').get('managed', []))
<class 'NoneType'>
```

The key is set, therefore `None` can still be returned instead of an empty list.  So, adding the code below would make a None value impossible while making Mypy happy.

```
        if managed is None:
            managed = []
```

The same idea would apply with `setdefault`.  That won't prevent a `None` if the key already exists.

Reply:
Also, more on this here: https://mypy.readthedocs.io/en/stable/kinds_of_types.html#strict-optional